Bug fix issue 30351

NewellClark opened the pull request:
This fixes [issue 30351](https://github.com/dotnet/runtime/issues/30351). I noticed there weren't any unit tests for System.CodeDom.IndentedTextWriter, so I went ahead and added those as well.

NewellClark committed:
Added tests for IndentedTextWriter

Added failing unit tests that demonstrate Issue #30351 with the IndentedTextWriter.

NewellClark committed:
IndentedTextWriter async overrides

Added overrides for async methods on IndentedTextWriter that call the respective async methods on the underlying TextWriter. 

Fix #30351

NewellClark committed:
Undid api modification in IndentedTextWriter

I can fix issue #30351 without modifying the API of IndentedTextWriter. 

Fix #30351

NewellClark committed:
Fix globalization issue with unit test

Apparently some of the CI tests run on a machine in a locale where commas are used as the decimal point. This caused some unit tests to fail. I've  changed those tests to use the invariant culture.

Fix issue #30351

buyaa-n commented on src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs:
```diff
@@ -141,6 +159,36 @@ public override void Write(string format, params object?[] arg)
             _writer.Write(format, arg);
         }
 
+        public override async Task WriteAsync(char value)
```
I believe all of this new public APIs also need to be added in ref https://github.com/dotnet/runtime/blob/8d7e074fe88d05a66f70e1717c1bea490cd41d2e/src/libraries/System.Runtime/ref/System.Runtime.cs#L5148-L5190

NewellClark replied:
Oh okay. I wasn't sure if that was necessary because they're virtual methods defined on a base class, which already has the refs. I'll go ahead and add them.

NewellClark replied:
Okay, I've added them.

buyaa-n commented on src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj:
```diff
@@ -98,18 +99,12 @@
-    <Compile Include="$(CommonTestPath)System\IO\TempFile.cs"
-             Link="Common\System\IO\TempFile.cs" />
-    <Compile Include="$(CommonTestPath)System\Runtime\Serialization\Formatters\BinaryFormatterHelpers.cs"
-             Link="Common\System\Runtime\Serialization\Formatters\BinaryFormatterHelpers.cs" />
```
Could you revert unwanted/unrelated changes, it would be cleaner for review

NewellClark replied:
My apologies. Not sure how that got changed. I guess VS changed them? I've reverted them.

buyaa-n replied:
No problem, thanks

buyaa-n commented on src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/IndentedTextWriterTests.cs:
```diff
@@ -0,0 +1,606 @@
+            await callWriteLineAsync(itw);
+
+            Assert.Equal(nameof(IndentedTextWriter.WriteLineAsync), indicator.LastCalledMethod);
+            Assert.Equal($"{expected}", indicator.GetStringBuilder().ToString());
```
```suggestion
            Assert.Equal(expected, indicator.GetStringBuilder().ToString());
```

NewellClark committed:
Update System.CodeDom.Tests.csproj

buyaa-n commented on src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/IndentedTextWriterTests.cs:
```diff
@@ -0,0 +1,606 @@
+            await callWriteLineAsync(itw);
+
+            Assert.Equal(nameof(IndentedTextWriter.WriteLineAsync), indicator.LastCalledMethod);
+            Assert.Equal($"{prefix}{NewLine}{TabString}{expected}", indicator.GetStringBuilder().ToString());
```
Could we add a test testing the original scenario, that shows that indentation is added after calling `IndentedTextWriter.WriteLineAsync();`, something like:
```csharp
        [Fact]
        public async Task TestIdentationAfterParameterlessWriteLineAsync()
        {
            var asyncSw = new StringWriter();
            var asyncTw = new IndentedTextWriter(asyncSw);

            asyncTw.Indent++;
            await asyncTw.WriteLineAsync("Foo");
            await asyncTw.WriteLineAsync("Bar");
            await asyncTw.WriteLineAsync();
            await asyncTw.WriteLineAsync("Quux");
            await asyncTw.WriteLineAsync("Wibble");

            var sw = new StringWriter();
            var tw = new IndentedTextWriter(sw);

            tw.Indent++;
            tw.WriteLine("Foo");
            tw.WriteLine("Bar");
            tw.WriteLine();
            tw.WriteLine("Quux");
            tw.WriteLine("Wibble");

            Assert.Equal(sw.GetStringBuilder().ToString(), asyncSw.GetStringBuilder().ToString());
        }
```

NewellClark committed:
Added overrides to ref library.

buyaa-n commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -5156,7 +5156,9 @@ public partial class IndentedTextWriter : System.IO.TextWriter
         [System.Diagnostics.CodeAnalysis.AllowNullAttribute]
         public override string NewLine { get { throw null; } set { } }
         public override void Close() { }
+        public override ValueTask DisposeAsync() => throw null;
```
Oops, types need to have fully qualified name in ref
```suggestion
        public override System.Threading.Tasks.ValueTask DisposeAsync() => throw null;
```

NewellClark replied:
Sorry about that. These "ref" projects are uncharted territory for me. I'll fix them and add the new unit tests.

buyaa-n replied:
No problem at all, thank you for working on it

NewellClark committed:
Made suggested edits

Corrected errors in refs.
Added a unit test for parameterless WriteLineAsync method. 
Cleaned up unit test with pointless interpolated string.

NewellClark commented:
/azp run runtime

buyaa-n commented on src/libraries/System.Runtime/System.Runtime.sln:
```diff
@@ -1,4 +1,8 @@
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio Version 16
+VisualStudioVersion = 16.0.30804.86
+MinimumVisualStudioVersion = 10.0.40219.1
```
Oops sorry, again VS changed this, please revert

NewellClark replied:
Reverted.

NewellClark committed:
Reverted unintended changes to System.Runtime.sln

VS made useless changes to System.Runtime.sln file.

buyaa-n commented:
Thank you @NewellClark, the PR looks good to me, i have 2 questions for @stephentoub and @carlossanlop 
1. Do we want to add the APIs into ref? i don't see any reason not to except it is not gone through API review
2. I am not sure if the tests should have added in System.IO, i am OK having it in `CodeDom` though, just saw there is existing test for `IndentedTextWriter` in System.IO

NewellClark commented:
> Thank you @NewellClark, the PR looks good to me, i have 2 questions for @stephentoub and @carlossanlop
> 
> 1. Do we want to add the APIs into ref? i don't see any reason not to except it is not gone through API review
> 2. I am not sure if the tests should have added in System.IO, i am OK having it in `CodeDom` though, just saw there is existing test for `IndentedTextWriter` in System.IO

I have just a few comments:
1) While we're discussing APIs, we may want to consider adding `protected virtual Task OutputTabsAsync()`, as well as `public Task WriteLineNoTabsAsync(string? s)`, to match the existing synchronous APIs that are specific to `IndentedTextWriter`. For example, if someone subclasses and overrides `OutputTabs()`, they should really also be overriding `OutputTabsAsync()`, otherwise they'll get the same bug in their subclass that we just fixed in `IndentedTextWriter`. 
2) Oops, My bad. I didn't see the existing tests in `System.IO`.

buyaa-n commented:
> I can submit an API review if you'd like.

Sorry for late response @NewellClark, I was in doubt if its need to go through API review, in general all APIs added to ref should, but it is just adding missing overloads not adding anything new. I talk with @stephentoub offline and confirmed we don't need API review here

> While we're discussing APIs, we may want to consider adding `protected virtual Task OutputTabsAsync()`, as well as `public Task WriteLineNoTabsAsync(string? s)`, to match the existing synchronous APIs that are specific to `IndentedTextWrite`. For example, if someone subclasses and overrides `OutputTabs()`, they should really also be overriding `OutputTabsAsync()`, otherwise they'll get the same bug in their subclass that we just fixed in `IndentedTextWriter`.

That is good point, those are also looks like missing implementation, as synchronous versions already in the ref https://github.com/dotnet/runtime/blob/167c598c71b7f77fe8bf5db2bb113c2565d5c84d/src/libraries/System.Runtime/ref/System.Runtime.cs#L5160 I agree we should add Async versions too

> Oops, My bad. I didn't see the existing tests in System.IO.

Yeah it is confusing, from namespace i would think it is related to `CodeDome`, i guess it is added to `IO` because it is a writer, we might want to move tests to `IO` @carlossanlop what you think?

NewellClark commented:
@buyaa-n Sounds good. I'll move the tests to `System.IO` and add the async versions of existing methods previously discussed.

NewellClark committed:
Added async versions of existing APIs

- Added OutputTabsAsync()
- Added WriteLineNoTabsAsync(string)

buyaa-n reviewed (approved):
Thank you very much @NewellClark, looks good to me, lets wait for System.IO area owner @carlossanlop's review/approval

carlossanlop commented on src/libraries/System.Runtime/ref/System.Runtime.cs:
```diff
@@ -5156,8 +5156,11 @@ public partial class IndentedTextWriter : System.IO.TextWriter
         [System.Diagnostics.CodeAnalysis.AllowNullAttribute]
         public override string NewLine { get { throw null; } set { } }
         public override void Close() { }
+        public override System.Threading.Tasks.ValueTask DisposeAsync() => throw null;
```
Here and elsewhere: I'm not sure if using `=>` would cause trouble, so let's stick to safety and use the same pattern used in all our refs (traditional parenthesis):
```suggestion
        public override System.Threading.Tasks.ValueTask DisposeAsync() { throw null; }
```

NewellClark replied:
Fixed

carlossanlop commented on src/libraries/System.IO/tests/IndentedTextWriter.cs:
```diff
@@ -178,5 +465,330 @@ public static async Task Writes_ProducesExpectedOutput(string newline)
+        private const string TabString = "   ";
+        private const string NewLine = "\n";
+
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework)]
```
Did you have to add this because the CI failing in .NET Framework? If you didn't see any failures, maybe the attribute is not needed.

NewellClark replied:
I just ran the tests without it and they passed. Not sure why I added them, it was a couple weeks ago. I'll remove them.

buyaa-n replied:
Well, CodeDome test project had .Net framework target `net461`, i guess it was failing before moving tests to System.IO https://github.com/dotnet/runtime/blob/32d664d10ce92f6535606b4ca658d0e9cfd754a4/src/libraries/System.CodeDom/tests/System.CodeDom.Tests.csproj#L4

now Syste.IO doesn't have .Net framework target

carlossanlop reviewed (changes requested):
Looks good. I left a few suggestions for you to consider. Please add documentation so it can be approved.

NewellClark committed:
Applied suggested changes

NewellClark commented on src/libraries/System.Private.CoreLib/src/System/CodeDom/Compiler/IndentedTextWriter.cs:
```diff
@@ -49,8 +51,12 @@ public int Indent
 
         public override void Close() => _writer.Close();
 
+        public override ValueTask DisposeAsync() => _writer.DisposeAsync();
```
I just ran the tests again and they passed without it, so I've gone ahead and removed it.

NewellClark commented:
@carlossanlop I've implemented your suggested changes.

buyaa-n commented:
Thank you @NewellClark, we can merge it now
